src/util.rs

```diff
@@ -24,6 +25,41 @@ use windows::{
     },
 };
 
+pub(crate) const fn win_guid_to_u128(guid: &GUID) -> u128 {
+    let data4_u64 = u64::from_be_bytes(guid.data4);
+    ((guid.data1 as u128) << 96) | ((guid.data2 as u128) << 80) | ((guid.data3 as u128) << 64) | (data4_u64 as u128)
+}
+
+pub(crate) fn win_pstr_to_string(pstr: ::windows_sys::core::PSTR) -> Result<String, Error> {
+    Ok(unsafe {
+        std::ffi::CStr::from_ptr(pstr as *const std::ffi::c_char)
+            .to_str()
+            .map_err(|e| format!("Invalid UTF-8 sequence: {}", e))
+    }?
+    .to_owned())
+}
+
+pub(crate) fn win_pwstr_to_string(pwstr: ::windows_sys::core::PWSTR) -> Result<String, Error> {
+    unsafe {
+        if pwstr.is_null() {
```

Comment:
Since this is a bit awkward to handle, and were making this unsafe anyway, I think the best way to express this is to force callers to handle the nullability, and add a safety comment saying that the pointer must never be null

Reply:
Also moving error handling to the call site allows us to return better errors. e.g: `FriendlyName field of adapter was unexpectedly null` vs `Null pointer received`. Additionally we can simply assert that the pointer is non-null at the call site if the docs specify that it should never be null.